subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -4204,102 +4193,77 @@ static void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx,
 		}
 
 		client_ctx->processed_req = NULL;
-
-		r = lwm2m_send_message(msg);
-		if (r < 0) {
-			LOG_ERR("Err sending response: %d", r);
-			lwm2m_reset_message(msg, true);
-		}
+		lwm2m_send_message_async(msg);
 	} else {
 		LOG_DBG("No handler for response");
 	}
 }
 
-static void retransmit_request(struct k_work *work)
+/* returns ms until the next retransmission is due, or INT32_MAX
+ * if no retransmissions are necessary
+ */
+static int32_t retransmit_request(struct lwm2m_ctx *client_ctx,
+				  const uint32_t timestamp)
 {
-	struct lwm2m_ctx *client_ctx;
 	struct lwm2m_message *msg;
-	struct coap_pending *pending;
-	int32_t remaining;
-	int ret;
-
-	client_ctx = CONTAINER_OF(work, struct lwm2m_ctx, retransmit_work);
-	pending = coap_pending_next_to_expire(client_ctx->pendings,
-					      CONFIG_LWM2M_ENGINE_MAX_PENDING);
-	if (!pending) {
-		return;
-	}
-
-	remaining = pending->t0 + pending->timeout - k_uptime_get_32();
-	if (remaining > 0) {
-		/* First message to expire was removed from the list,
-		 * schedule next.
-		 */
-		goto next;
-	}
-
-	msg = find_msg(pending, NULL);
-	if (!msg) {
-		LOG_ERR("pending has no valid LwM2M message!");
-		coap_pending_clear(pending);
-		goto next;
-	}
+	struct coap_pending *p;
+	uint32_t expiry, remaining = INT32_MAX;
+	int i;
 
-	if (!coap_pending_cycle(pending)) {
-		/* pending request has expired */
-		if (msg->message_timeout_cb) {
-			msg->message_timeout_cb(msg);
+	for (i = 0, p = client_ctx->pendings;
+	     i < CONFIG_LWM2M_ENGINE_MAX_PENDING; i++, p++) {
+		if (!p->timeout) {
+			continue;
 		}
 
-		/*
-		 * coap_pending_clear() is called in lwm2m_reset_message()
-		 * which balances the ref we made in coap_pending_cycle()
-		 */
-		lwm2m_reset_message(msg, true);
-		goto next;
-	}
-
-	if (msg->acknowledged) {
-		/* No need to retransmit, just keep the timer running to
-		 * timeout in case no response arrives.
-		 */
-		goto next;
-	}
-
-	LOG_INF("Resending message: %p", msg);
-	msg->send_attempts++;
-
-	ret = sys_mutex_lock(&client_ctx->send_lock, K_FOREVER);
-	__ASSERT(ret == 0, "sys_mutex_lock failed with %d", ret);
-
-	if (msg->ctx == NULL) {
-		LOG_INF("Response for %p already handled", msg);
-		goto next_locked;
-	}
+		expiry = p->t0 + p->timeout;
+		if (expiry < timestamp) {
```

Comment:
You should rather check if `((int32_t)(expiry - timestamp) < 0)`, otherwise it won't work on uint32_t overflow (i. e. when the `expiry` is near to UINT32_MAX, but `timestamp` has already overflown.

Reply:
Good point.